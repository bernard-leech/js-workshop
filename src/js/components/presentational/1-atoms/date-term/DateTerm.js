// @flow
import React from "react"
import { format as formatFunc, differenceInCalendarDays } from 'date-fns'

export const getRelativeDay = (date, relativeDate = new Date()) => {
  const isRelativeDateANumber = (relativeDate instanceof Date)

  const dateDifference = isRelativeDateANumber ? differenceInCalendarDays(date, relativeDate) : '0'

  let output = `${dateDifference} days left`

  if (dateDifference === 0) {
    output = 'Due today'
  }

  if (dateDifference < 0) {
    output = `${Math.abs(dateDifference)} days late`
  }

  return output
}

function DateTerm(props) {
  const { date, format = 'D MMM YYYY', isRelative } = props
  
  const dateToDisplay = isRelative ? getRelativeDay(date) : formatFunc(date, format)

  return (
    <span className="date-container">{dateToDisplay}</span>
  )
}

export default DateTerm
