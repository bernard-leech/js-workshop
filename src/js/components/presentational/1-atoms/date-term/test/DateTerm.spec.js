import React from 'react'
import TestRenderer from 'react-test-renderer'

import { props } from '../test/DateTerm.mock'

import DateTerm, { getRelativeDay } from '../DateTerm'
import { max } from 'date-fns';

const dueTodayText = 'Due today'

it('renders correctly', () => {
  const tree = TestRenderer
    .create(<DateTerm {...props} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test(`getRelativeDay() returns ${dueTodayText} when provided with todays date`, () => {
  const myDateTerm = new Date()
  expect(getRelativeDay(myDateTerm)).toBe(dueTodayText)
})

test(`getRelativeDay() returns 1 day left`, () => {
  const { date } = props

  const myDate = new Date(date)
  let yesterday = new Date(date)
  yesterday.setDate(myDate.getDate() - 1)

  expect(getRelativeDay(date, yesterday)).toBe("1 days left")
})

test(`getRelativeDay() returns "not a number", when provided with a string `, () => {
  const { date } = props
  expect(getRelativeDay(date, 'string')).toBe("0 days left")
})
