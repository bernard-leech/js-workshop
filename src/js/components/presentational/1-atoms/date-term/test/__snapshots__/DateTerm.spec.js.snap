// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<span
  className="date-container"
>
  4 Feb 2019
</span>
`;
